### PR TITLE
Lfs bootloader

### PR DESCRIPTION
**Pull request by jlaitine**
This add support for NOR flash to bootloader

### CODE REVIEW

File: platforms/nuttx/src/bootloader/microchip/mpfs/main.c [hunk 1]
```diff
@@ -182,9 +212,10 @@ board_deinit(void)
 	//	putreg32(RCC_AHB1RSTR_OTGFSRST, STM32_RCC_AHB1RSTR);
 #endif
 
-#ifdef MMCSD_IMAGE_FN
-	close(mmcsd_fd);
-	umount("/sdcard");
+#ifdef IMAGE_FN
+	close(fd);
+	umount("/fs/lfs");
```
Comment (jlaitine) [inline platforms/nuttx/src/bootloader/microchip/mpfs/main.c:217]: this was leftovers, need to still remove

File: platforms/nuttx/src/bootloader/microchip/mpfs/main.c [hunk 2]
```diff
@@ -459,32 +538,70 @@ bootloader_main(void)
...
 
-			do {
-				got += read(mmcsd_fd, (void *)(APP_LOAD_ADDRESS + got), 512);
-			} while (!(got % 512));
+		unsigned reads_left = image_sz / FLASH_RW_BLOCK - 1;
 
-		} else {
-			/* Mount succeeded, but file not found. Make sure the "boot" directory is there */
-			ret = mkdir("/sdcard/boot", S_IRWXU | S_IRWXG | S_IRWXO);
+		// read the rest in 32k blocks
```
Comment (jlaitine) [inline platforms/nuttx/src/bootloader/microchip/mpfs/main.c:585]: // read the rest in FLASH_RW_BLOCK size blocks

File: platforms/nuttx/src/bootloader/microchip/mpfs/main.c [hunk 3]
```diff
@@ -585,6 +702,29 @@ bootloader_main(void)
...
+
+			/* Write the residue */
+			unsigned bytes = end_address - first_unwritten;
+			unsigned n_blocks = bytes / 256;
+
+			if (bytes % 256) {
+				n_blocks += 1;
+			}
+
+			flash_write_pages(first_unwritten / 256, n_blocks,
```
Comment (jlaitine) [inline platforms/nuttx/src/bootloader/microchip/mpfs/main.c:719]: if (n_blocks) {

File: boards/ssrc/saluki/src/CMakeLists.txt [hunk 4]
```diff
@@ -63,6 +64,7 @@ else()
 		led.c
 		toc.c
 		mpfs_emmcsd.c
+		mpfs_spinor.c
```
Comment (jpaali) [inline boards/ssrc/saluki/src/CMakeLists.txt:67]: This mpfs_spinor.c is not included into the commit

**Review by jpaali — COMMENTED** [review]
One file missing otherwise ready for merge

File: boards/ssrc/saluki/src/CMakeLists.txt (hunk 4 above)
Comment (jlaitine) [reply boards/ssrc/saluki/src/CMakeLists.txt:67]: oops, thanks. added the file

File: boards/ssrc/saluki/src/mpfs_spinor.c [hunk 5]
```diff
@@ -0,0 +1,105 @@
+/****************************************************************************
+ * boards/risc-v/mpfs/icicle/src/mpfs_emmcsd.c
```
Comment (jlaitine) [inline boards/ssrc/saluki/src/mpfs_spinor.c:2]: Need to go through these file headers under board, for proper license text. this is coped from nuttx, and is probably not the right one for PX4

File: boards/ssrc/saluki/src/mpfs_spinor.c [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ *
+ * Description:
+ *   Configure the SPI NOR flash driver.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; A negated errno value is returned
+ *   to indicate the nature of any failure.
+ *
+ ****************************************************************************/
+int mksmartfs(FAR const char *pathname, uint16_t sectorsize);
```
Comment (jpaali) [inline boards/ssrc/saluki/src/mpfs_spinor.c:52]: we don't probably need this?

File: boards/ssrc/saluki/src/mpfs_spinor.c (hunk 5 above)
Comment (jpaali) [reply boards/ssrc/saluki/src/mpfs_spinor.c:2]: ok

**Review by jpaali — APPROVED** [review]

**Pull request merged**